docs/decisions/0011-depending-on-multiple-event-bus-implementations.rst

```diff
@@ -0,0 +1,72 @@
+
+Decision
+********
+
+We will depend on *all* known and supported event bus implementations in every IDA that needs to produce to or consume from the event bus.
+
+All deployers will have all implementations installed and available for use. Dependency conflicts will need to be addressed as they arise during development and maintenance.
+
+Consequences
+************
+
+By keeping all implementations in the base requirements, we have the following benefits:
+
+- Version management is overall simplified. By including package versions in source control, we can more easily maintain stable development and production environments. Deployers can collaborate more easily because they are using shared version specifications.
+- Developers will not have to repeatedly adjust their devstack containers to include the desired event bus implementation.
+- Dependency conflicts will be discovered relatively quickly. If an IDA and an event bus implementation have incompatible constraints on a transitive dependency, there is no risk that we will upgrade the dependency in the IDA and then write code that relies on new features; we will instead have to solve the conflict up front.
+- As a corollary, we can readily switch from explicit requirements to out-of-tree requirements, since dependencies will be kept consistent. This would not be true the other way around, since we might have to contend with various conflicts.
+- This approach does not preclude the use of other implementations that are not part of the Open edX project. If a community member creates their own implementation and it is not offered to or accepted into the Open edX project, it can still be installed out-of-tree by that deployer.
+
+There are also some downsides:
+
+- The dependency tree is larger, and is more or less guaranteed to include at least one top-level dependency that each deployer is not using. Not only does each dependency take up bandwidth to download, space on disk, and time spent in upgrading packages, but it can also bring licensing and architecture incompatibilities.
+- More concretely, event-bus-kafka depends on confluent-kafka, but that package does not provide binaries compatible with Linux on Apple Silicon. This means that developers on Apple M1 laptops cannot install this package in devstack at all. So as a knock-on effect, event-bus-kafka has had to keep confluent-kafka out of its own in-tree dependencies; deployers wishing to use it have to install and version that package out-of-tree anyhow. This obviates a number of the stated benefits for development and version management, at least for the Kafka users.
```

Comment:
Presumably this will be fixed at some point (there are PRs open for it already upstream), and then the benefits of this approach will be restored, right?

Reply:
True! I'll make a note that it might be a temporary situation. (I'm not holding my breath, though -- the librdkafka folks don't seem interested in finding an alternative CI provider that supports Apple Silicon.)